### PR TITLE
Python runtime rework 

### PR DESCRIPTION
**Pull request by cehongwang**
# Unify Python and C++ TensorRT runtimes

PythonTorchTensorRTModule is removed. Both runtimes now live behind the same TorchTensorRTModule, with a single use_python_runtime flag (on CompilationSettings) selecting which path to take at engine setup.

What's changed
One module, two backends. TorchTensorRTModule.setup_engine() constructs either torch.classes.tensorrt.Engine (C++) or TRTEngine (Python) and binds the matching op to self.execute_engine_op. forward() just calls the bound op — no per-iteration branching, no separate module class.
Two equivalent ops. tensorrt::execute_engine (C++) and tensorrt::execute_engine_python (Python) are both registered and have the same signature, fake kernel, and semantics. The Python op is registered unconditionally so it's available even when the C++ runtime is loaded.

# Checklist:

- [ ] My code follows the style guidelines of this project (You can use the linters)
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas and hacks
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests to verify my fix or my feature
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have added the relevant labels to my PR in so that relevant reviewers are notified

### CODE REVIEW

File: core/runtime/register_jit_hooks.cpp [hunk 1]
```diff
@@ -101,54 +101,14 @@ static auto TORCHTRT_UNUSED TRTEngineTSRegistrtion =
                           : TRTEngine::ResourceAllocationStrategy::kStatic);
             })
         .def_readwrite("use_pre_allocated_outputs", &TRTEngine::use_pre_allocated_outputs)
+        .def_readwrite("pre_allocated_outputs", &TRTEngine::pre_allocated_outputs)
```
Comment (narendasan) [inline core/runtime/register_jit_hooks.cpp:51]: Why do we need both?

File: docsrc/tutorials/deployment/distributed_inference.rst [hunk 2]
```diff
@@ -86,531 +74,66 @@ inference in ``distributed_context`` for safe NCCL lifecycle management:
...
-* The warmup forward pass should use ``torch._dynamo.mark_dynamic()`` to match the
-  generate loop and avoid a recompile.
-* **Non-default TP subgroup** — if the TRT engine should use a subgroup communicator
-  (e.g. tensor-parallel inside a data-parallel job), pass the subgroup instead of
-  ``dist.group.WORLD``:
-
-  .. code-block:: python
-
-      tp_group = dist.new_group(ranks=[0, 1])
-      with torch_tensorrt.distributed.distributed_context(tp_group, trt_model) as model:
```
Comment (narendasan) [inline docsrc/tutorials/deployment/distributed_inference.rst:155]: Are we keeping this or removing it?

File: py/torch_tensorrt/dynamo/runtime/_serialized_engine_layout.py [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+REQUIRES_OUTPUT_ALLOCATOR_IDX = 9
+RESOURCE_ALLOCATION_STRATEGY_IDX = 10
+SERIALIZATION_LEN = 11
+
+SERIALIZED_ENGINE_BINDING_DELIM = "%"
+SERIALIZED_RT_DEVICE_DELIM = "%"
+
+# (torch.ops.tensorrt name, module global holding the expected value, normalizer)
+_LayoutCheck = Tuple[str, str, Callable[[Any], Any]]
+_LAYOUT_CPP_CHECKS: tuple[_LayoutCheck, ...] = (
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/runtime/_serialized_engine_layout.py:62]: What do we use this for?

Comment (narendasan) [reply py/torch_tensorrt/dynamo/runtime/_serialized_engine_layout.py:62]: Can you add comments about what its is going for and how to extend the ABI?

File: py/torch_tensorrt/dynamo/_exporter.py [hunk 4]
```diff
@@ -557,12 +562,25 @@ def replace_execute_engine_no_op_node(
...
-            trt_engine = torch.classes.tensorrt.Engine(tuple(packed_engine_info))
+            metadata = TorchTensorRTModule.decode_metadata(
+                packed_engine_info[SERIALIZED_METADATA_IDX]
+            )
+            use_python_runtime = getattr(
+                metadata.get("settings"), "use_python_runtime", False
+            )
+            if use_python_runtime:
+                from torch_tensorrt.dynamo.runtime._TRTEngine import TRTEngine
+
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/_exporter.py:559]: Is this going to work without the upstream python changes?

File: scripts/remove_set_runtime_backend_with.py [hunk 5]
```diff
@@ -0,0 +1,99 @@
+#!/usr/bin/env python3
+"""Remove obsolete ``with *.runtime.set_runtime_backend(...):`` blocks and dedent bodies."""
```
Comment (narendasan) [inline scripts/remove_set_runtime_backend_with.py]: What is this file for?

**Review by narendasan — COMMENTED** [review]
Not sure this PR is clean yet

File: core/runtime/register_jit_hooks.cpp (hunk 1 above)
Comment (cehongwang) [reply core/runtime/register_jit_hooks.cpp:51]: The first one is the flag, and the second one is the tensor. There is a test that compares the output tensor pointer to verify it is the same tensor. So the second one is needed

**Review by zewenli98 — COMMENTED** [review]
Is this PR based on #4164? Why do we need two separate PRs? 
Can you complete `Description` of the PRs so that other folks can understand why and what the PRs are doing?

File: py/torch_tensorrt/dynamo/_exporter.py (hunk 4 above)
Comment (cehongwang) [reply py/torch_tensorrt/dynamo/_exporter.py:559]: Oh it actually not after I restore the PyTorch changes. Changing it back to throwing an error

File: py/torch_tensorrt/dynamo/runtime/_serialized_engine_layout.py (hunk 3 above)
Comment (cehongwang) [reply py/torch_tensorrt/dynamo/runtime/_serialized_engine_layout.py:62]: This is the compatibility check that you mentioned before. In a full build, when importing torch_tensorrt, we perform a check on whether the C++ indices are the same as Python here. I will add a comment

**Comment by cehongwang** [comment]
I will revert the commit `168bdea547a614b26c256527f65217e2d0fbf222` once the pytorch changes are landed.

File: core/runtime/TRTEngine.h [hunk 6]
```diff
@@ -181,7 +181,6 @@ struct TRTEngine : torch::CustomClassHolder {
   int64_t get_streamable_device_memory_budget();
   int64_t get_automatic_device_memory_budget();
   std::vector<at::Tensor> infer_outputs(std::vector<std::vector<int64_t>> input_shapes);
-  void set_pre_allocated_outputs(bool enable);
```
Comment (narendasan) [inline core/runtime/TRTEngine.h:186]: Not needed anymore?

File: py/torch_tensorrt/dynamo/conversion/_conversion.py [hunk 7]
```diff
@@ -379,7 +369,7 @@ def convert_module(
                 "For TRT-LLM fallback, set TRTLLM_PLUGINS_PATH or USE_TRTLLM_PLUGINS=1."
             )
 
-    return rt_cls(
+    return TorchTensorRTModule(
         serialized_engine=serialized_interpreter_result.serialized_engine,
         input_binding_names=list(serialized_interpreter_result.input_names),
         output_binding_names=list(serialized_interpreter_result.output_names),
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/conversion/_conversion.py:375]: Are we dropping use_python_runtime then?

File: py/torch_tensorrt/dynamo/runtime/meta_ops/register_meta_ops.py [hunk 8]
```diff
@@ -320,6 +320,7 @@ def no_op_placeholder_for_execute_engine(
     serialized_metadata: str,
     serialized_target_platform: str,
     serialized_require_output_allocator: str,
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/runtime/meta_ops/register_meta_ops.py:321]: Could you also add the md_trt tag too, I might have forgotten

File: py/torch_tensorrt/dynamo/runtime/_serialized_engine_layout.py [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+
+import tensorrt as trt
+import torch
+import torch_tensorrt
+from torch_tensorrt._features import ENABLED_FEATURES
+
+ABI_VERSION = "9"
+
+
+class SerializedInfoIndex(IntEnum):
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/runtime/_serialized_engine_layout.py:22]: Feat. request for later is to move this into some sort of data that both C++ and python can load at build time

File: py/torch_tensorrt/dynamo/runtime/_TorchTensorRTModule.py [hunk 10]
```diff
@@ -367,39 +378,62 @@ def set_extra_state(self, state: SerializedTorchTensorRTModuleFmt) -> None:
...
             self.weight_name_map = metadata["weight_name_map"]
-            self.output_tensors_are_unowned = metadata["output_tensors_are_unowned"]
             self.symbolic_shape_expressions = metadata["inout_symexprs"]
-            self.engine.set_output_tensors_as_unowned(self.output_tensors_are_unowned)
 
+            # Re-resolve the runtime now that we have the loaded settings: the
+            # original __init__ kwarg may have been False, but a saved engine
+            # can still pin use_python_runtime=True via the settings blob.
+            self._use_python_runtime = (
+                getattr(self.settings, "use_python_runtime", False)
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/runtime/_TorchTensorRTModule.py:387]: Ah ok

**Review by narendasan — APPROVED** [review]
Cool I think this is looking good. One thing I think this shows is maybe we can keep the use_python_runtime but only register with torch in python only? then we arent losing user flexibility

File: docsrc/py_api/runtime.rst [hunk 11]
```diff
@@ -27,13 +27,20 @@ Functions
 
 .. autofunction:: enable_output_allocator
 
+Runtime backend
+---------------
+
+Execution uses the C++ runtime engine when it is installed in the build; otherwise the
+Python runtime engine is used. There is no separate process-wide backend switch
+in ``torch_tensorrt.runtime``.
```
Comment (zewenli98) [inline docsrc/py_api/runtime.rst:35]: If c++ is built, is it able to switch to python runtime?

File: py/torch_tensorrt/dynamo/conversion/_conversion.py [hunk 12]
```diff
@@ -343,22 +342,13 @@ def convert_module(
...
-
-    if ENABLED_FEATURES.torch_tensorrt_runtime and not settings.use_python_runtime:
-        from torch_tensorrt.dynamo.runtime import TorchTensorRTModule
-
-        rt_cls = TorchTensorRTModule
-
-    elif (
-        not ENABLED_FEATURES.torch_tensorrt_runtime and not settings.use_python_runtime
-    ):
+    if not ENABLED_FEATURES.torch_tensorrt_runtime:
```
Comment (zewenli98) [inline py/torch_tensorrt/dynamo/conversion/_conversion.py:351]: Do you mean `torch_tensorrt_cpp_runtime`? the name is kind of confusing

**Review by zewenli98 — APPROVED** [review]
overall LGTM

File: core/runtime/TRTEngine.h (hunk 6 above)
Comment (zewenli98) [inline core/runtime/TRTEngine.h:186]: why is it removed?

Comment (cehongwang) [reply core/runtime/TRTEngine.h:186]: It's never used

File: docsrc/py_api/runtime.rst (hunk 11 above)
Comment (cehongwang) [reply docsrc/py_api/runtime.rst:35]: No

File: py/torch_tensorrt/dynamo/conversion/_conversion.py (hunk 12 above)
Comment (cehongwang) [reply py/torch_tensorrt/dynamo/conversion/_conversion.py:351]: I think torch_tensorrt_runtime is the cpp runtime. I did not invent the name. It's been there for a while and I am just using it
